dlp/deid.py

```diff
@@ -0,0 +1,548 @@
+# [END reidentify_with_fpe]
+
+
+# [START deidentify_with_date_shift]
+def deidentify_with_date_shift(parent, input_csv_file=None,
+                               output_csv_file=None, date_fields=None,
+                               lower_bound_days=None, upper_bound_days=None,
+                               context_field_id=None, wrapped_key=None,
+                               key_name=None):
+    """Uses the Data Loss Prevention API to deidentify dates in a CSV file by
+        pseudorandomly shifting them.
+    Args:
+        input_csv_file: The path to the CSV file to deidentify. The first row
+            of the file must specify column names, and all other rows must
+            contain valid values.
+        output_csv_file: The path to save the date-shifted CSV file.
+        date_fields: The list of (date) fields in the CSV file to date shift.
+            Example: ['birth_date', 'register_date']
+        lower_bound_days: The maximum number of days to shift a date backward
+        upper_bound_days: The maximum number of days to shift a date forward
+        context_field_id: (Optional) The column to determine date shift amount
+            based on. If this is not specified, a random shift amount will be
+            used for every row. If this is specified, then 'wrappedKey' and
+            'keyName' must also be set. Example:
+            contextFieldId = [{ 'name': 'user_id' }]
+        key_name: (Optional) The name of the Cloud KMS key used to encrypt
+            ('wrap') the AES-256 key. Example:
+            key_name = 'projects/YOUR_GCLOUD_PROJECT/locations/YOUR_LOCATION/
+            keyRings/YOUR_KEYRING_NAME/cryptoKeys/YOUR_KEY_NAME'
+        wrapped_key: (Optional) The encrypted ('wrapped') AES-256 key to use.
+            This key should be encrypted using the Cloud KMS key specified by
+            key_name.
+    Returns:
+        None; the response from the API is printed to the terminal.
+    """
+    # Import the client library
+    import google.cloud.dlp
+
+    # Instantiate a client
+    google.cloud.dlp.DlpServiceClient.SERVICE_ADDRESS = 'autopush-dlp.sandbox.googleapis.com' # DO NOT SUBMIT
+    dlp = google.cloud.dlp.DlpServiceClient()
+
+    # Add parent
+    parent = dlp.project_path(parent)
+
+    # Convert date field list to Protobuf type
+    def map_fields(field):
+        return {'name': field}
+
+    if date_fields:
+        date_fields = map(map_fields, date_fields)
+    else:
+        date_fields = []
+
+    # Read and parse the CSV file
+    import csv
+    from datetime import datetime
+    file = []
+    with open(input_csv_file, 'rb') as csvfile:
+        reader = csv.reader(csvfile)
+        for row in reader:
+            file.append(row)
+
+    #  Helper function for converting CSV rows to Protobuf types
+    def map_headers(header):
+        return {'name': header}
+
+    def map_data(value):
+        try:
+            date = datetime.strptime(value, '%m/%d/%Y')
```

Comment:
Hmm, the Node version will automatically try to guess the date format, but this Python function requires the format to be predetermined.

I'd rather not add dateutil to the requirements, though, since it's an arbitrary third-party library. Let's ask around about this tomorrow.

Reply:
Let's go ahead and use dateutil. Add it to requirements.txt in this directory and then we can use it to parse dynamically like Node.js samples do.